internal/api/provider/oidc.go

```diff
@@ -200,6 +204,127 @@ func parseLinkedinIDToken(token *oidc.IDToken) (*oidc.IDToken, *UserProvidedData
+}
+
+func (c *AzureIDTokenClaims) IsEmailVerified() bool {
+	// If xms_edov is not set, and an email is present or xms_edov is true,
+	// only then is the email regarded as verified.
+	// https://learn.microsoft.com/en-us/azure/active-directory/develop/migrate-off-email-claim-authorization#using-the-xms_edov-optional-claim-to-determine-email-verification-status-and-migrate-users
+	emailVerified := false
+
+	edov := c.XMicrosoftEmailDomainOwnerVerified
+
+	if edov == nil {
+		// An email is provided, but xms_edov is not -- probably not
+		// configured, so we must assume the email is verified as Azure
+		// will only send out a potentially unverified email address in
+		// single-tenanat apps.
+		emailVerified = c.Email != ""
+	} else {
+		edovBool := false
+
+		// Azure can't be trusted with how they encode the xms_edov
+		// claim. Sometimes it's "xms_edov": "1", sometimes "xms_edov": true.
+		switch edov.(type) {
+		case bool:
+			edovBool = edov.(bool)
+
+		case string:
+			edovBool = edov.(string) == "1" || edov.(string) == "true"
+
+		case int:
+			edovBool = edov.(int) != 0
+
+		case int32:
+			edovBool = edov.(int32) != 0
+
+		case int64:
+			edovBool = edov.(int64) != 0
+
+		case float32:
+			edovBool = edov.(float32) != 0
+
+		case float64:
+			edovBool = edov.(float64) != 0
+		}
+
+		emailVerified = c.Email != "" && edovBool
+	}
+
+	return emailVerified
+}
+
+// removeAzureClaimsFromCustomClaims contains the list of claims to be removed
+// from the CustomClaims map. See:
+// https://learn.microsoft.com/en-us/azure/active-directory/develop/id-token-claims-reference
+var removeAzureClaimsFromCustomClaims = []string{
+	"aud",
+	"iss",
+	"iat",
+	"nbf",
+	"exp",
+	"c_hash",
+	"at_hash",
+	"aio",
+	"nonce",
+	"rh",
+	"uti",
+	"jti",
+	"ver",
+	"sub",
+	"name",
+	"preferred_username",
+}
+
+func parseAzureIDToken(token *oidc.IDToken) (*oidc.IDToken, *UserProvidedData, error) {
+	var data UserProvidedData
+
+	var azureClaims AzureIDTokenClaims
+	if err := token.Claims(&azureClaims); err != nil {
+		return nil, nil, err
+	}
+
+	data.Metadata = &Claims{
+		Issuer:            token.Issuer,
+		Subject:           token.Subject,
+		Email:             azureClaims.Email,
+		EmailVerified:     azureClaims.IsEmailVerified(),
+		ProviderId:        token.Subject,
+		PreferredUsername: azureClaims.PreferredUsername,
+		CustomClaims:      make(map[string]any),
+	}
+
+	if data.Metadata.Email != "" {
+		data.Emails = append(data.Emails, Email{
+			Email:    data.Metadata.Email,
+			Verified: data.Metadata.EmailVerified,
+			Primary:  true,
+		})
+	}
+
+	if err := token.Claims(&data.Metadata.CustomClaims); err != nil {
+		return nil, nil, err
+	}
+
+	if data.Metadata.CustomClaims != nil {
+		for _, claim := range removeAzureClaimsFromCustomClaims {
+			delete(data.Metadata.CustomClaims, claim)
+		}
+	}
+
+	if len(data.Emails) <= 0 {
+		return nil, nil, fmt.Errorf("provider: Azure OIDC ID token from issuer %q must contain an email address", token.Issuer)
```

Comment:
maybe we can reword this to:
```suggestion
		return nil, nil, fmt.Errorf("provider: Azure OIDC ID token from issuer %q must contain a valid email address", token.Issuer)
```